clang/lib/CodeGen/CodeGenModule.cpp

```diff
@@ -6235,6 +6235,20 @@ void CodeGenModule::EmitGlobalVarDefinition(const VarDecl *D,
     getCUDARuntime().handleVarRegistration(D, *GV);
   }
 
+  if (LangOpts.SYCLIsDevice) {
+    const RecordDecl *RD = D->getType()->getAsRecordDecl();
+    if (RD && RD->hasAttr<SYCLDeviceGlobalAttr>()) {
+      // SYCL device globals are initialized externally
+      GV->setExternallyInitialized(true);
```

Comment:
@smanna12 do you think we can set this universally for all `device_globals`. In NVPTX/AMDGCN this is correct but I'm not entirely sure if we should consider device globals for SPIR-V targets to be externally initialized.

Reply:
What is the change to the generated IR? I believe we attempted to achieve this by marking the `device_global` in `llvm.compiler.used`, which seems to avoid the optimizer optimizing it when it looks like it won't be initialized. I am not sure why that doesn't reach the CUDA and HIP path though.

If this works with SPIR-V too, I am curious as to whether we could drop the `llvm.compiler.used` marking for `device_global`.

Reply:
It turns out that `file-table-tform` is getting rid of the `llvm.compiler.used` symbol. But this is before the `llc` step which also involves `llvm-linking` libclc and libdevice for some reason

Reply:
I think I agree with that assessment. Does it work with SPIR-V targets as well? If so, @asudarsa did we have some cleanup of the global variable somewhere that we won't have to do anymore?

Reply:
From https://llvm.org/docs/LangRef.html#the-llvm-compiler-used-global-variable:
"This is a rare construct that should only be used in rare circumstances, and should not be exposed to source languages."

Why is this being used instead of https://llvm.org/docs/LangRef.html#the-llvm-used-global-variable

Reply:
I don't know why this was used instead of `llvm.used`. I have changed this to a more portable solution based on linkage and renaming static device global symbols in order to *hide* them, despite having external linkage. I think this is better than relying on compiler vars like `llvm.compiler.used` or indeed `llvm.used`

Reply:
I think @AaronBallman 's  point was that the solution might be as simple as using `llvm.used` instead of `llvm.compiler.used` because `llvm.used` is supposed to provide the functionality you are attempting to implement here via changing linkage. Changing the Linkage is a bit strange and we're not sure if that is the right move here.

Reply:
That's an easier fix. I've made a PR [here](https://github.com/intel/llvm/pull/15212)

Reply:
So are you saying that regardless of binary semantics, emulating internal linkage by mangling names that on the ELF level are VIS_DEFAULT at the language level works "as if" they were invisible to other compilation units and thus a valid implementation of C and C++'s linkage rules for `static` globals?

Reply:
I am indeed saying that. We don’t need to worry about leakage of symbols across the host boundary as symbols are still internally linked in host compilation. Mangled symbols persist if we are to link device modules together, however since we mangle device symbols on a per module basis, there is no chance of symbol corruption when linking device module A with device module B, each containing the same static device global symbol in source code.

Reply:
:+1: Gotcha. I think I'm happy with that for device code. I agree it might be nice to use `llvm.used` or similar, but there are too many built in assumptions around how the linkage model is implemented in clang and llvm *in general* to make me comfortable that such a marker will be respected *in general* - and I think your experiments are showing that it is indeed brittle.

Reply:
> That actually doesn't work. llvm.used instead of llvm.compiler.used is still getting removed at file-table-tform stage, which results in the symbol being tampered with/removed at a later step for NVPTX targets. 

That is a bug with file-table-tform; it's not honoring the documented LLVM IR expectations. IMO, the compiler should not be working around this bug.

> Also bear in mind that the global variable is only marked to have external linkage for the device compilation pass, not for host compilation, so host behaviour is not changed. Since the variable is externally initialised as far as the device compilation is concerned, it makes sense to make it externally linked.

The symbol is marked with internal linkage and we're generating a new symbol with external linkage as a result; that this doesn't happen on host doesn't mean this is reasonable behavior -- nobody is going to expect external symbols to be generated for internal linkage.

Reply:
The removal of device globals from `llvm.compiler.used` is actually happening deliberately at [`sycl-post-link`](https://github.com/intel/llvm/blob/sycl/llvm/tools/sycl-post-link/sycl-post-link.cpp#L582).

This is a feature, not a bug. It breaks because the `sycl-post-link` step is followed by a `llc` type step which includes linking bitcode modules with `-mlink-builtin-bitcode`. The `llvm-link` pipeline ends up removing the global symbol which is no longer protected by `llvm.compiler.used`.

Reply:
I've made a patch here to modify `sycl-post-link` behaviour [here](https://github.com/intel/llvm/pull/15224). I think this should solve the problem

Reply:
> The removal of device globals from llvm.compiler.used is actually happening deliberately at [sycl-post-link](https://github.com/intel/llvm/blob/sycl/llvm/tools/sycl-post-link/sycl-post-link.cpp#L582), not file-table-tform. Apologies for the misdirection.

No worries, but let's make sure we're on the same page: `llvm.compiler.used` is expected to be dropped by tools, so that behavior is reasonable. `llvm.used` is expected to *not* be dropped because it needs to survive to the final linking phase. So my suggestion is for us to use `llvm.used` from Clang instead of `llvm.compiler.used`, and then fix the tooling that was dropping `llvm.used`. Does that make sense?

Reply:
I see, thanks for explanation. I've changed use of `llvm.compiler.used` to `llvm.used` for static device globals in the  PR [here](https://github.com/intel/llvm/pull/15224).

Reply:
As far as I'm concerned, the ideal fix would be to have all SYCL backends (NVPTX, AMDGN and SPIR-V) treat `llvm.compiler.used` in the same way, by removing the global symbol at bc -> asm lowering. This is what NVPTX does currently and I think it is sensible behaviour. In this way we can make sure the symbol lasts for all IR parts of the pipeline. For device code we don't need to worry about post lowering linking optimisations that might remove the symbol, at least we don't currently need the symbol at this stage using the `sycl-post-link` method, so I think `llvm.compiler.used` is preferable to `llvm.used`.

In this way we would stop removing `llvm.compiler.used` at `sycl-post-link`, which is unintuitive behaviour especially given that it happens before further LLVM passes are run (not honouring `llvm.compiler.used`, which is supposed to last until lowering).

@AaronBallman is that a solution you can get behind?

Reply:
I think I can get behind that, yeah. My biggest concern is regarding changing the linkage; so long as we don't need to do that, I'll be happy!